Sources/MapboxDirections/MatrixOptions.swift

```diff
@@ -0,0 +1,200 @@
+        waypoints = try container.decode([Waypoint].self, forKey: .waypoints)
+        profileIdentifier = try container.decode(ProfileIdentifier.self, forKey: .profileIdentifier)
+        if let annotations = try container.decodeIfPresent(String.self, forKey: .attributeOptions) {
+            attributeOptions = Set<AttributeOption>(from: annotations)
+        }
+        destinations = try container.decodeIfPresent([Int].self, forKey: .destinations)
+        sources = try container.decodeIfPresent([Int].self, forKey: .sources)
+    }
+    
+    // MARK: Getting the Request URL
+    
+    internal var coordinates: String? {
+        waypoints.map {
+            $0.coordinate.requestDescription
+        }.joined(separator: ";")
+    }
+    
+    /**
+        An array of URL query items to include in an HTTP request.
+    */
+   var abridgedPath: String {
+       return "directions-matrix/v1/\(profileIdentifier.rawValue)"
+   }
+
+   /**
+    The path of the request URL, not including the hostname or any parameters.
+    */
+   var path: String {
+    guard let coordinates = coordinates,
+          !coordinates.isEmpty else {
+        assertionFailure("No query")
+        return ""
+    }
+    return "\(abridgedPath)/\(coordinates)"
+   }
+    
+    /**
+     An array of URL query items (parameters) to include in an HTTP request.
+     */
+    public var urlQueryItems: [URLQueryItem] {
+        var queryItems: [URLQueryItem] = []
+        
+        let annotations = self.attributeOptions
+        if !annotations.isEmpty && annotations.isSubset(of: [.distance, .travelTime]) {
+            queryItems.append((URLQueryItem(name: "annotations", value: annotations.queryDescription)))
+        }
+        
+        let mustArriveOnDrivingSide = !waypoints.filter { !$0.allowsArrivingOnOppositeSide }.isEmpty
+        if mustArriveOnDrivingSide {
+            let approaches = waypoints.map { $0.allowsArrivingOnOppositeSide ? "unrestricted" : "curb" }
+            queryItems.append(URLQueryItem(name: "approaches", value: approaches.joined(separator: ";")))
+        }
+        
+        if let destinations = self.destinations {
+            let destinationStrings = destinations.map { String($0)}
+            queryItems.append(URLQueryItem(name: "destinations", value: destinationStrings.joined(separator: ";")))
+        }
+        
+        if let sources = self.sources {
+            let sourceStrings = sources.map { String($0)}
+            queryItems.append(URLQueryItem(name: "sources", value: sourceStrings.joined(separator: ";")))
+        }
+        return queryItems
+    }
+}
+
+extension MatrixOptions {
+    
+    /**
+     Attributes are metadata information for requesting matricies.
+     
+     Options selected affect the contents of the resulting `MatrixResponse`.
+     */
+    public typealias AttributeOptions = Set<AttributeOption>
+    
+    public struct AttributeOption: Hashable, RawRepresentable {
```

Comment:
I assume this is a separate type than the main `AttributeOptions` because the Matrix API only supports a limited number of `annotations` values. However, some of the `AttributeOptions` values are already limited to one profile or another, so I don’t think it would be entirely surprising if we were to reuse that type but document some values like `congestionLevel` as applying only to the Directions and Map Matching APIs.

The other difference seems to be that `Matrix.AttributeOptions` is a `Set` rather than an `OptionSet`. In hindsight, this would’ve been a pretty good move for `AttributeOptions` in v1.0 or v2.0, but I wonder if adopting that approach only for the Matrix API would prove confusing.

Reply:
Updated original AttributeOptions docs and re-used them for Matrix.